source/mir/blas/kernel/gemm.d

```diff
@@ -0,0 +1,151 @@
+}
+body
+{
+    pragma(inline, false);
+    enum msg = "Wrong kernel compile time arguments.";
+    static assert(conj == Conj.none && P == 1 || conj != Conj.none && P == 2, msg);
+    do
+    {
+        V[P][N][M] reg = void;
+        foreach (m; Iota!M)
+        foreach (n; Iota!N)
+        foreach (p; Iota!P)
+            static if (add)
+                reg[m][n][p] = c[0][m][n][p];
+            else
+                reg[m][n][p] = 0;
+        size_t i = columns;
+        do
+        {
+            V[P][N] ai = void;
+            V[P][M] bi = void;
+            foreach (n; Iota!N)
+            foreach (p; Iota!P)
+                ai[n][p] = a[0][n][p];
+            foreach (m; Iota!M)
+            foreach (p; Iota!P)
+                bi[m][p] = b[0][m][p];
+            a++;
+            b++;
+            foreach (m; Iota!M)
+            foreach (n; Iota!N)
+            {
+                reg[m][n][0] += ai[n][0] * bi[m][0];
+                static if (conj == Conj.complexNone)
+                {
+                    reg[m][n][1] += ai[n][0] * bi[m][1];
+                    reg[m][n][0] -= ai[n][1] * bi[m][1];
+                    reg[m][n][1] += ai[n][1] * bi[m][0];
+                }
+                else static if (conj == Conj.complexA)
+                {
+                    reg[m][n][1] += ai[n][0] * bi[m][1];
+                    reg[m][n][0] += ai[n][1] * bi[m][1];
+                    reg[m][n][1] -= ai[n][1] * bi[m][0];
+                }
+                else static if (conj == Conj.complexB)
+                {
+                    reg[m][n][1] -= ai[n][0] * bi[m][1];
+                    reg[m][n][0] += ai[n][1] * bi[m][1];
+                    reg[m][n][1] += ai[n][1] * bi[m][0];
+                }
+                else static if (conj == Conj.complexC)
+                {
+                    reg[m][n][1] -= ai[n][0] * bi[m][1];
+                    reg[m][n][0] -= ai[n][1] * bi[m][1];
+                    reg[m][n][1] -= ai[n][1] * bi[m][0];
+                }
+                else static assert(conj == Conj.none, msg);
+            }
+        }
+        while (--i);
+        b -= columns;
+        foreach (m; Iota!M)
+        foreach (n; Iota!N)
+        foreach (p; Iota!P)
+            c[0][m][n][p] = reg[m][n][p];
+        c++;
+    }
+    while (--rows);
+}
+
+unittest
+{
+    import std.meta: AliasSeq;
+    import std.typecons: Yes, No;
+    with(Conj)
+    foreach (conj; AliasSeq!(none, complexNone, complexA, complexB))
+    foreach (add; AliasSeq!(No.add, Yes.add))
+    {
+        enum P = conj == none ? 1 : 2;
+        {alias temp = gemmMicroKernel!(conj, add, P, 2 / P, 4 / P, float, float);}
+        {alias temp = gemmMicroKernel!(conj, add, P, 2 / P, 4 / P, double, double);}
+        version(LDC)
+        {
+            {alias temp = gemmMicroKernel!(conj, add, P, 2 / P, 4 / P, __vector(float[8]), float);}
+            {alias temp = gemmMicroKernel!(conj, add, P, 2 / P, 4 / P, __vector(double[4]), double);}
+        }
```

Comment:
Can we have an assert statement here to verify the kernel code is actually correct?